### PR TITLE
[TT-Train] Fix GCC build: qualify self-referential using declarations (#37922)

### LINKED ISSUE #37922: [TT-Train]: GCC build failure — 5 device operation headers use self-referential using declarations
### Component / Area

TT-Train / Metal Ops / Device Operations

### Issue Type (optional)

Other

### Observed

tt-train fails to compile with GCC (any version) due to self-referential `using` type alias declarations in 5 device operation headers. The pattern `using operation_attributes_t = operation_attributes_t;` inside a struct creates a member name that shadows the namespace-scope type per C++ standard [basic.scope.class]/2. GCC correctly rejects this:

```
error: declaration of 'using operation_attributes_t = struct ttml::metal::ops::layernorm_fw::device::operation_attributes_t'
       changes meaning of 'operation_attributes_t' [-fpermissive]
```

Clang accepts this code (more lenient name lookup), but GCC 12, 13, and 14 all reject it.

### Expected

tt-train should compile with both Clang and GCC, consistent with tt-metal's own compiler support (Clang 17+ and GCC 12+, with toolchain files for both).

### 1. Steps (exact commands)

```bash
# Build tt-metal with clang-20 first (succeeds)
source python_env/bin/activate
./build_metal.sh --debug --build-all --enable-ccache
deactivate

# Attempt standalone tt-train build with GCC 12
cd tt-train
rm -rf build/
cmake -DCMAKE_BUILD_TYPE=Debug \
      -DTT_TRAIN_INHERIT_COMPILER=OFF \
      -DCMAKE_C_COMPILER=gcc-12 \
      -DCMAKE_CXX_COMPILER=g++-12 \
      -B build -GNinja
cmake --build build --config Debug
# Fails with "changes meaning of" errors in 5 headers
```

### 2. Input data / link or description

No input data needed — build-time issue.

**Broken pattern** (5 files — unqualified, self-referential):
```cpp
namespace ttml::metal::ops::layernorm_fw::device {
struct LayerNormForwardDeviceOperation {
    using operation_attributes_t = operation_attributes_t;      // BUG: self-referential
    using tensor_args_t = tensor_args_t;
    using spec_return_value_t = spec_return_value_t;
    using tensor_return_value_t = tensor_return_value_t;
};
}
```

**Correct pattern** (8 other files in the same codebase — fully qualified):
```cpp
namespace ttml::metal::ops::rmsnorm_fw::device {
struct RMSNormForwardDeviceOperation {
    using operation_attributes_t = ttml::metal::ops::rmsnorm_fw::device::operation_attributes_t;
    using tensor_args_t = ttml::metal::ops::rmsnorm_fw::device::tensor_args_t;
    using spec_return_value_t = ttml::metal::ops::rmsnorm_fw::device::spec_return_value_t;
    using tensor_return_value_t = ttml::metal::ops::rmsnorm_fw::device::tensor_return_value_t;
};
}
```

**Affected files (5):**

| File | Namespace |
|------|-----------|
| `tt-train/sources/ttml/metal/ops/layernorm_fw/device/layernorm_fw_device_operation.hpp` | `ttml::metal::ops::layernorm_fw::device` |
| `tt-train/sources/ttml/metal/ops/layernorm_bw/device/layernorm_bw_device_operation.hpp` | `ttml::metal::ops::layernorm_bw::device` |
| `tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_device_operation.hpp` | `ttml::metal::ops::swiglu_fw::device` |
| `tt-train/sources/ttml/metal/optimizers/adamw/device/adamw_device_operation.hpp` | `ttml::metal::optimizers::adamw::device` |
| `tt-train/sources/ttml/metal/optimizers/sgd_fused/device/sgd_fused_device_operation.hpp` | `ttml::metal::optimizers::sgd_fused::device` |

**Already correct files (8):** rmsnorm_fw, rmsnorm_bw, softmax, silu_bw, cross_entropy_fw, cross_entropy_bw, sdpa_fw, profiler_no_op.

### 3. Frequency

Always — deterministic build failure with any GCC version.

### 1. Software Versions

- GCC 12.3.0 (tested in Docker container), affects all GCC versions
- Clang 17+ accepts the code
- tt-metal main at commit `26bdadf292` (2026-02-15)
- Ubuntu 22.04

### 2. Hardware Details

N/A — build-time issue, not hardware-dependent.

### Is this a regression?

No

### Regression Details

The unqualified pattern was introduced when these 5 operations were added. It was not caught because tt-train CI only builds with Clang. The 8 other operation files were written correctly from the start.

### Logs & Diagnostics

Full error output for `layernorm_fw` (same pattern repeats for all 5 files):
```
tt-train/sources/ttml/metal/ops/layernorm_fw/device/layernorm_fw_device_operation.hpp:16:11:
  error: declaration of 'using operation_attributes_t = struct ttml::metal::ops::layernorm_fw::device::operation_attributes_t'
         changes meaning of 'operation_attributes_t' [-fpermissive]
   16 |     using operation_attributes_t = operation_attributes_t;
      |           ^~~~~~~~~~~~~~~~~~~~~
tt-train/sources/ttml/metal/ops/layernorm_fw/device/layernorm_fw_device_operation_types.hpp:15:8:
  note: 'operation_attributes_t' declared here as 'struct ttml::metal::ops::layernorm_fw::device::operation_attributes_t'
   15 | struct operation_attributes_t {
      |        ^~~~~~~~~~~~~~~~~~~~~
```

Same error repeats for `tensor_args_t`, `spec_return_value_t`, and `tensor_return_value_t` in each of the 5 files.

### Priority

P3

### Impact

Prevents building tt-train with GCC. The fix is trivial — add full namespace qualification to the `using` declarations in 5 files, matching the pattern already used by 8 other operations in the same codebase (one-line change per type alias, 20 lines total).

### PR DESCRIPTION
**Pull request by epam-iaroslav-voitovych**
### Ticket
Fixes #37922

### Problem description
5 tt-train device operation headers use self-referential `using` type alias declarations:
```cpp
using operation_attributes_t = operation_attributes_t;  // unqualified — self-referential
```
This violates C++ [basic.scope.class]/2: the member name on the left shadows the namespace-scope type on the right, making the declaration self-referential. GCC (all versions) correctly rejects this with `-fpermissive`; Clang is lenient and accepts it.

8 other device operations in the same codebase already use the correct fully-qualified pattern.

### What's changed
Added full namespace qualification to the `using` declarations in the 5 affected files, matching the pattern already used by rmsnorm_fw, rmsnorm_bw, softmax, silu_bw, cross_entropy_fw, cross_entropy_bw, sdpa_fw, and profiler_no_op.

#### Before (broken with GCC)
```cpp
namespace ttml::metal::ops::layernorm_fw::device {
struct LayerNormForwardDeviceOperation {
    using operation_attributes_t = operation_attributes_t;
};
}
```

#### After (works with both GCC and Clang)
```cpp
namespace ttml::metal::ops::layernorm_fw::device {
struct LayerNormForwardDeviceOperation {
    using operation_attributes_t = ttml::metal::ops::layernorm_fw::device::operation_attributes_t;
};
}
```

#### Files changed (5 files, 20 lines)
- `tt-train/sources/ttml/metal/ops/layernorm_fw/device/layernorm_fw_device_operation.hpp`
- `tt-train/sources/ttml/metal/ops/layernorm_bw/device/layernorm_bw_device_operation.hpp`
- `tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_device_operation.hpp`
- `tt-train/sources/ttml/metal/optimizers/adamw/device/adamw_device_operation.hpp`
- `tt-train/sources/ttml/metal/optimizers/sgd_fused/device/sgd_fused_device_operation.hpp`

#### Verification
Container build with `--tt-train-compiler gcc-12` merging this fix + PR #37166 (compiler selection) on top of main — verifies tt-train compiles successfully with GCC 12.

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:ivoitovych/issue-37922-fix-gcc-self-referential-using)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ivoitovych/issue-37922-fix-gcc-self-referential-using)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ivoitovych/issue-37922-fix-gcc-self-referential-using)
- [ ] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ivoitovych/issue-37922-fix-gcc-self-referential-using)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ivoitovych/issue-37922-fix-gcc-self-referential-using)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ivoitovych/issue-37922-fix-gcc-self-referential-using)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

**Comment by epam-iaroslav-voitovych** [comment]
#### Container verification

Verified in Docker container builds using [parameterized build scripts](https://github.com/ivoitovych/tt-container/tree/feature/parameterized-build) that merge branches on top of main and optionally override the tt-train compiler:

| # | Test | Compiler | Branches merged | Result |
|---|------|----------|-----------------|--------|
| 1 | gcc-12 override | gcc-12 | This PR + PR #37166 (compiler selection) | **PASSED** |
| 2 | Default (no override) | clang-20 (inherited from CMakeCache.txt) | This PR + PR #37166 | **PASSED** |

Build 1 confirms GCC 12 now compiles tt-train successfully with the qualified `using` declarations.
Build 2 confirms no regression on the default Clang path.

Build commands used:
```bash
# Build 1: gcc-12 override
./build_tt_docker.sh --branch main \
  --merge-branch ivoitovych/issue-36993-fix-gcc-constexpr-lambda \
  --merge-branch ivoitovych/tt-metal:ivoitovych/issue-37922-fix-gcc-self-referential-using \
  --tt-train-compiler gcc-12

# Build 2: default compiler (clang-20 inherited from tt-metal CMakeCache.txt)
./build_tt_docker.sh --branch main \
  --merge-branch ivoitovych/issue-36993-fix-gcc-constexpr-lambda \
  --merge-branch ivoitovych/tt-metal:ivoitovych/issue-37922-fix-gcc-self-referential-using
```

**Review by mdragulaTT — APPROVED** [review]
I'm okay with this. Thank you @epam-iaroslav-voitovych 👍🏼

**Review by nsextonTT — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]
ideally those structures should be called 

```
LayerNormBackwardAttributes
LayerNormBackwardTensorArgs
...
```
